memory-db/src/lib.rs

```diff
@@ -160,19 +160,151 @@ impl<H, KF, T> Eq for MemoryDB<H, KF, T>
 pub trait KeyFunction<H: KeyHasher> {
 	type Key: Send + Sync + Clone + hash::Hash + Eq;
 
-	fn key(hash: &H::Out, prefix: &[u8]) -> Self::Key;
+	fn key(hash: &H::Out, prefix: Prefix) -> Self::Key;
+}
+
+
+/// Make database key from hash and prefix.
+///
+/// Prefix could be ordered, but key hash appending will break this property.
+/// If true ordering is needed we would need a fix length encoding for prefix
+/// (so a maximum length and additional byte (or chunk of prefix encoded)).
+pub fn prefixed_key<H: KeyHasher>(key: &H::Out, prefix: Prefix) -> Vec<u8> {
+	let mut prefixed_key = Vec::with_capacity(key.as_ref().len() + prefix.0.len() + 1);
```

Comment:
I removed the nibble length : my reasoning for its need was wrong: two nodes using this prefix (one with padding and one without are in the same trie path so if both got the same hash it would indicate a trie loop or a hash collision).